### ISSUE TITLE
Deprecation notices for PSR-4 autoloading with correct paths?

### ISSUE THREAD

**Issue by ondrejmirtes** [composer/composer#8683]
My `composer.json`:

```json
	"autoload": {
		"psr-4": {
			"Slevomat\\Phing\\": [
				"build/Phing"
			],
			"Slevomat\\UI\\": [
				"app/ui"
			],
			"Slevomat\\": [
				"app/model",
				"app/models",
				"app/services"
			]
		}
	},
	"autoload-dev": {
		"psr-4": {
			"Slevomat\\UI\\": [
				"tests/ui"
			],
			"Slevomat\\PHPStan\\": [
				"build/PHPStan",
				"tests/Build/PHPStan"
			],
			"Slevomat\\": [
				"tests/model",
				"tests/models",
				"tests/services",
				"tests/stubs",
				"tests"
			]
		},
		"files": [
			"vendor/slevomat/coding-standard/SlevomatCodingStandard/Helpers/TokenHelper.php"
		]
	}
```

Output of `composer diagnose`:

```
Checking composer.json: WARNING
The package "chrisboulton/php-diff" is pointing to a commit-ref, this is bad practice and can cause unforeseen issues.
The package "consistence-php/consistence" is pointing to a commit-ref, this is bad practice and can cause unforeseen issues.
The package "genkgo/camt" is pointing to a commit-ref, this is bad practice and can cause unforeseen issues.
The package "kdyby/forms-replicator" is pointing to a commit-ref, this is bad practice and can cause unforeseen issues.
require.aws/aws-php-sns-message-validator : exact version constraints (1.6.0) should be avoided if the package follows semantic versioning
require.aws/aws-sdk-php : exact version constraints (3.133.23) should be avoided if the package follows semantic versioning
require.beberlei/doctrineextensions : exact version constraints (1.2.6) should be avoided if the package follows semantic versioning
require.chrisboulton/php-diff : unbound version constraints (dev-master#784ff85) should be avoided
require.clue/buzz-react : exact version constraints (2.7.0) should be avoided if the package follows semantic versioning
require.composer/ca-bundle : exact version constraints (1.2.6) should be avoided if the package follows semantic versioning
require.composer/semver : exact version constraints (1.5.1) should be avoided if the package follows semantic versioning
require.consistence-php/consistence : unbound version constraints (dev-master#f2af962) should be avoided
require.consistence/consistence : exact version constraints (2.0.1) should be avoided if the package follows semantic versioning
require.doctrine/annotations : exact version constraints (1.8.0) should be avoided if the package follows semantic versioning
require.doctrine/cache : exact version constraints (1.10.0) should be avoided if the package follows semantic versioning
require.doctrine/collections : exact version constraints (1.6.4) should be avoided if the package follows semantic versioning
require.doctrine/common : exact version constraints (2.12.0) should be avoided if the package follows semantic versioning
require.doctrine/dbal : exact version constraints (2.10.1) should be avoided if the package follows semantic versioning
require.doctrine/event-manager : exact version constraints (1.1.0) should be avoided if the package follows semantic versioning
require.doctrine/orm : exact version constraints (2.7.0.1) should be avoided if the package follows semantic versioning
require.doctrine/persistence : exact version constraints (1.3.6) should be avoided if the package follows semantic versioning
require.doctrine/reflection : exact version constraints (1.1.0) should be avoided if the package follows semantic versioning
require.elasticsearch/elasticsearch : exact version constraints (6.7.1) should be avoided if the package follows semantic versioning
require.endroid/qr-code : exact version constraints (3.7.5) should be avoided if the package follows semantic versioning
require.genkgo/camt : unbound version constraints (dev-master#8890566 as 0.4.99) should be avoided
require.geoip2/geoip2 : exact version constraints (2.10.0) should be avoided if the package follows semantic versioning
require.giggsey/libphonenumber-for-php : exact version constraints (8.11.5) should be avoided if the package follows semantic versioning
require.google/apiclient : exact version constraints (2.4.0) should be avoided if the package follows semantic versioning
require.guzzlehttp/guzzle : exact version constraints (6.5.2) should be avoided if the package follows semantic versioning
require.guzzlehttp/promises : exact version constraints (1.3.1) should be avoided if the package follows semantic versioning
require.guzzlehttp/psr7 : exact version constraints (1.6.1) should be avoided if the package follows semantic versioning
require.jyxo/php : exact version constraints (5.0.2) should be avoided if the package follows semantic versioning
require.latte/latte : exact version constraints (2.7.0) should be avoided if the package follows semantic versioning
require.minishlink/web-push : exact version constraints (5.2.4) should be avoided if the package follows semantic versioning
require.mjaschen/phpgeo : exact version constraints (3.0.0) should be avoided if the package follows semantic versioning
require.mobiledetect/mobiledetectlib : exact version constraints (2.8.34) should be avoided if the package follows semantic versioning
require.monolog/monolog : exact version constraints (2.0.2) should be avoided if the package follows semantic versioning
require.nette/application : exact version constraints (3.0.4) should be avoided if the package follows semantic versioning
require.nette/bootstrap : exact version constraints (3.0.1) should be avoided if the package follows semantic versioning
require.nette/caching : exact version constraints (3.0.1) should be avoided if the package follows semantic versioning
require.nette/component-model : exact version constraints (3.0.0) should be avoided if the package follows semantic versioning
require.nette/di : exact version constraints (3.0.1) should be avoided if the package follows semantic versioning
require.nette/finder : exact version constraints (2.5.2) should be avoided if the package follows semantic versioning
require.nette/forms : exact version constraints (3.0.3) should be avoided if the package follows semantic versioning
require.nette/http : exact version constraints (3.0.3) should be avoided if the package follows semantic versioning
require.nette/mail : exact version constraints (3.1.0) should be avoided if the package follows semantic versioning
require.nette/neon : exact version constraints (3.1.2) should be avoided if the package follows semantic versioning
require.nette/php-generator : exact version constraints (3.3.4) should be avoided if the package follows semantic versioning
require.nette/routing : exact version constraints (3.0.0) should be avoided if the package follows semantic versioning
require.nette/schema : exact version constraints (1.0.2) should be avoided if the package follows semantic versioning
require.nette/security : exact version constraints (3.0.2) should be avoided if the package follows semantic versioning
require.nette/utils : exact version constraints (3.1.1) should be avoided if the package follows semantic versioning
require.paragonie/constant_time_encoding : exact version constraints (2.3.0) should be avoided if the package follows semantic versioning
require.phing/phing : exact version constraints (2.16.3) should be avoided if the package follows semantic versioning
require.phpoffice/phpspreadsheet : exact version constraints (1.11.0) should be avoided if the package follows semantic versioning
require.pixel418/markdownify : exact version constraints (2.3.1) should be avoided if the package follows semantic versioning
require.pragmarx/google2fa : exact version constraints (7.0.0) should be avoided if the package follows semantic versioning
require.predis/predis : exact version constraints (1.1.1) should be avoided if the package follows semantic versioning
require.psr/http-message : exact version constraints (1.0.1) should be avoided if the package follows semantic versioning
require.pyyoshi/shortuuid-php : exact version constraints (1.0.7) should be avoided if the package follows semantic versioning
require.ramsey/uuid : exact version constraints (3.9.3) should be avoided if the package follows semantic versioning
require.ramsey/uuid-doctrine : exact version constraints (1.6.0) should be avoided if the package follows semantic versioning
require.react/dns : exact version constraints (1.2.0) should be avoided if the package follows semantic versioning
require.react/event-loop : exact version constraints (1.1.1) should be avoided if the package follows semantic versioning
require.react/promise : exact version constraints (2.7.1) should be avoided if the package follows semantic versioning
require.react/promise-timer : exact version constraints (1.5.1) should be avoided if the package follows semantic versioning
require.react/socket : exact version constraints (1.3.0) should be avoided if the package follows semantic versioning
require.ronanguilloux/isocodes : exact version constraints (2.2.3) should be avoided if the package follows semantic versioning
require.ruflin/elastica : exact version constraints (6.1.1) should be avoided if the package follows semantic versioning
require.scienta/doctrine-json-functions : exact version constraints (4.1.2) should be avoided if the package follows semantic versioning
require.slevomat/csob-gateway : exact version constraints (5.0.0) should be avoided if the package follows semantic versioning
require.slevomat/eet-client : exact version constraints (2.0.0) should be avoided if the package follows semantic versioning
require.symfony/console : exact version constraints (5.0.5) should be avoided if the package follows semantic versioning
require.symfony/css-selector : exact version constraints (5.0.5) should be avoided if the package follows semantic versioning
require.symfony/dom-crawler : exact version constraints (5.0.5) should be avoided if the package follows semantic versioning
require.symfony/event-dispatcher : exact version constraints (5.0.5) should be avoided if the package follows semantic versioning
require.symfony/finder : exact version constraints (5.0.5) should be avoided if the package follows semantic versioning
require.tecnickcom/tc-lib-barcode : exact version constraints (1.15.20) should be avoided if the package follows semantic versioning
require.tecnickcom/tc-lib-color : exact version constraints (1.12.15) should be avoided if the package follows semantic versioning
require.texy/texy : exact version constraints (3.1.1) should be avoided if the package follows semantic versioning
require.tracy/tracy : exact version constraints (2.7.3) should be avoided if the package follows semantic versioning
require.ua-parser/uap-php : exact version constraints (3.9.7) should be avoided if the package follows semantic versioning
require.vasek-purchart/phing-symfony-command-task : exact version constraints (3.0) should be avoided if the package follows semantic versioning
require.vipnytt/sitemapparser : exact version constraints (1.1.0) should be avoided if the package follows semantic versioning
require.web-token/jwt-checker : exact version constraints (v1.3.9) should be avoided if the package follows semantic versioning
require.web-token/jwt-core : exact version constraints (v1.3.9) should be avoided if the package follows semantic versioning
require.web-token/jwt-signature : exact version constraints (v1.3.9) should be avoided if the package follows semantic versioning
require.web-token/jwt-signature-algorithm-rsa : exact version constraints (v1.3.9) should be avoided if the package follows semantic versioning
require.wyrihaximus/react-guzzle-psr7 : exact version constraints (2.1.2) should be avoided if the package follows semantic versioning
require.zbateson/mail-mime-parser : exact version constraints (1.2.0) should be avoided if the package follows semantic versioning
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.10.0
PHP version: 7.4.2
PHP binary path: /usr/local/Cellar/php/7.4.2/bin/php
OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019

```

When I run this command:

```
composer dump-autoload
```

I get the following output:

```
Deprecation Notice: Class Slevomat\ExchangeRate\ExchangeRateServiceTest located in ./tests/services/ExchangeRate/ExchangeRateServiceTest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0.
```

I don't see the error between path configuration and the file path to the class. Can you point it out? :) I suspect this might be a Composer bug of an unexpected but valid configuration, for example the same namespaces used between `autoload`/`autoload-dev` or something like that. WDYT?

**Comment by Seldaek** [comment]
Is the repo public?

But anyway, I have a feeling it is failing when loading `autoload-dev => Slevomat\\ => tests`, as that will go over classes including those mapped by tests/services. So it kinda looks like a misconfiguration as you have multiple paths within each other serving as the same namespace root. I do see that it will work, but it's also a mess. 

I'd say either normalize your classes to have a single namespace-root, or avoid putting stuff in tests/ directory (I mean `tests/*/*.php` is fine, but `tests/*.php` is not as it requires adding this conflicting rule.), or last option is to just use a `classmap: ["tests/"]` rule to autoload all this.. But then you don't benefit from psr-4 advantages anymore and will have to redump autoloader when adding a new class.

**Comment by rob006** [comment]
The same error is reported for `yiisoft/yii2-queue`.

https://github.com/yiisoft/yii2-queue/blob/2818e2ce348a166d52d1d0f3fcee83401b4fe106/composer.json#L46-L58

What exactly in this config does not comply with psr-4?

**Comment by Seldaek** [comment]
It's the same problem, `"yii\\queue\\": "src",` points to a folder which has other PSR-4 roots within it, like `"yii\\queue\\amqp\\": "src/drivers/amqp",` so when scanning src/ for yii\queue namespace it finds things in src/drivers/amqp/... which do not match the expected `yii\queue\drivers\amqp\...` class name.

Maybe would be fixable on composer side by processing things in the most-specific to least-specific order, and then avoiding scanning the same files twice, that hopefully would remove such conflicts.

**Comment by ondrejmirtes** [comment]
> avoiding scanning the same files twice, that hopefully would remove such conflicts

I agree, the goal should be to bring the number of deprecation notices down in case there's no real problem because the class will be autoloaded thanks to a different rule in `autoload` section.

For example, [`consistence/consistence`](https://github.com/consistence/consistence) looks like this:

```
	"autoload": {
		"psr-4": { "Consistence\\": [ "src" ] },
		"classmap": [ "src" ]
	},
```

It's because there are some classes in `src` that do not adhere to PSR-4, like `Consistence\Math\NonNegativeIntegerExpectedException`, located in: `src/Math/exceptions/NonNegativeIntegerExpectedException.php`.

Even if it will be loaded thanks to `classmap`, Composer 1.10 outputs this message:

> Deprecation Notice: Class Consistence\Math\NonNegativeIntegerExpectedException located in ./vendor/consistence/consistence/src/Math/exceptions/NonNegativeIntegerExpectedException.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar://~/.bin/composer/src/Composer/Autoload/ClassMapGenerator.php:185

**Comment by Seldaek** [comment]
@ondrejmirtes that's an even worse case as it's not covered by another psr-4 rule but a whole different rule. We can try to also cover this case, but really for a package named consistence, I'd expect a less inconsistent codebase.. There is no reason to put these exception classes in a random subfolder. Or if you do then there should be an `..\Exceptions\..` namespace too.. Anyway I see it's not your package but I wonder what people are thinking sometimes.

**Comment by Seldaek** [comment]
Alrighty, wasn't all too bad after all the fix, and it might even speed the dumping up a little bit for those with lots of overlapping rules.

Please try with `composer self-upate --snapshot` if it fixes it for you, or if you suddenly can't load classes because I messed up ;)

**Comment by VasekPurchart** [comment]
@Seldaek Hi, it is my codebase and if you are interested I can provide the original reasoning behind this:

* when you have a codebase with *a lot* of exceptions (named exceptions for all usecases, not just reusing the same few exceptions in the same package/namespace/whole world) the directory structure becomes very unreadable
* putting exceptions in a separate namespace
  1) makes no sense, since they semantically belong into the same namespace
  2) also makes it inconvenient for work with the classes (they are in the same namespace for a reason)
  3) is imho a bad practice overall (the same as all `Controllers`, `Entities` etc. from Symfony), namespaces should not be "directories"

By the way exceptions are not the only case - if you have some classes only for test it is very similar and again, it makes no sense to put it in a separate namespace.

I was not happy with this solution either, but it kind of stood the test of time and caused very little problems. You just have to run build if you encounter this, but since there are more reasons to do it in more complex setups it is not an important issue.

If you know about another solution, which would not hamper with the namespaces, then I would be glad for advice. At the point when I was researching the possibilities the only other solution would be to generate the list of these files, which did not remove the build step + was awful in other ways.

So in the end it is consistent, just not PSR-4 only. Kind of like using tabs instead of spaces, I know I am a rebel... :)

Anyway thanks for the quick response :)

**Comment by ondrejmirtes** [comment]
I confirm the snapshot version fixes this :) It now only reports a few cases where the file path is really wrong. The only reported 3rd party package is now:

> Deprecation Notice: Class Zxing\Common\CharacterSetEci\AbstractEnum\AbstractEnum located in ./vendor/khanamiryan/qrcode-detector-decoder/lib/Common/AbstractEnum.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar://~/.bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201

Thank you very much!

**Comment by ondrejmirtes** [comment]
I've also ran PHPStan after dumping the autoloader because it relies on it heavily and I can confirm the autoloading isn't broken.

**Comment by Seldaek** [comment]
Great thanks for the update. 

@VasekPurchart I get your point absolutely, namespaces as directories and especially as "things" like exceptions/controllers/.. does not make a ton of sense. But what you do is just not achievable with PSR-4 alone. I'd say maybe simply using classmap for all classes would be cleaner, but in practice your solution works, so I suppose it's fine. Just had to rant a little because people abusing things leads to more work for me, but in the end it was a simple fix so all good.

I'll try and get a patch release out soon.

**Comment by Seldaek** [comment]
Released now as 1.10.1
